File: clients/ts/src/policy.ts
import { err, ok } from "./errors.js";
import { request } from "./http.js";
import type { HttpContext, Policy, Result } from "./types.js";

/**
 * Namespace for reading the access control policy. Requires the admin role
 * (the configured `admin_role`, `"admin"` by default).
 *
 * The policy is defined in the server's settings directory (`policies.json`,
 * with every referenced role declared in `roles.json`); files are the only
 * write path. Edit the files and the server re-adopts them on change, on
 * SIGHUP, or on `wh.settings.reload()` (POST /v1/ops/settings/reload).
 * Validate a draft with the server's `wavehouse validate` command before
 * writing it to the directory.
 */
export class PolicyNamespace {
  private readonly _ctx: HttpContext;

  constructor(ctx: HttpContext) {
    this._ctx = ctx;
  }

  /** Get the current access control policy. */
  async get(opts?: { signal?: AbortSignal }): Promise<Result<Policy>> {
    const { data, error } = await request<Policy>(this._ctx, {
      method: "GET",
      path: "/v1/ops/policy",
      signal: opts?.signal,
    });
    if (error) return err(error);
    return ok(data!);
  }
}

File: clients/go/policy.go
package wavehouse

import (
	"context"
	"fmt"
)

// PolicyNamespace provides admin-only access-control policy management.
type PolicyNamespace struct {
	ctx httpContext
}

// Get returns the current access-control policy. Admin-only.
func (p *PolicyNamespace) Get(ctx context.Context) (*Policy, error) {
	var pol Policy
	if err := doRequest(ctx, p.ctx, requestOptions{
		method: "GET",
		path:   "/v1/ops/policy",
	}, &pol); err != nil {
		return nil, fmt.Errorf("get policy: %w", err)
	}
	return &pol, nil
}

// Set replaces the entire access-control policy. Admin-only.
func (p *PolicyNamespace) Set(ctx context.Context, pol *Policy) error {
	if err := doRequest(ctx, p.ctx, requestOptions{
		method: "PUT",
		path:   "/v1/ops/policy",
		body:   pol,
	}, nil); err != nil {
		return fmt.Errorf("set policy: %w", err)
	}
	return nil
}

// Validate checks a policy without applying it (dry run). Admin-only.
func (p *PolicyNamespace) Validate(ctx context.Context, pol *Policy) (*ValidationResult, error) {
	var result ValidationResult
	if err := doRequest(ctx, p.ctx, requestOptions{
		method: "POST",
		path:   "/v1/ops/policy/validate",
		body:   pol,
	}, &result); err != nil {
		return nil, fmt.Errorf("validate policy: %w", err)
	}
	return &result, nil
}
